### DISCUSSION TITLE
Existing devices not getting deleted

### DISCUSSION THREAD

**Discussion by sounak-1989** [edgexfoundry/community#111] [General]
Hey guys,

I am having a problem. Once I register some devices to core-metadata, even if I stop the device service or do a docker volume prune the devices still persist inside core-metadata. 

This is an issue because even if I change my device profile I am unable to remove the old profile that lingers there. I have to send a delete command for each and every device to clean up the downed devices.

Is there a solution to this problem? I have not seen this issue before updating to Kamakura.

**Comment by cloudxxx8** [comment]
It's not an issue, becuase it works as desgin.  core-metadata is the source of truth.  You have to manage devices and device profiles through the core-metadata APIs.  The pre-defined devices and device profiles in the Device services are just initializtion convenience for developers.
https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-metadata/2.3.0

**Reply by sounak-1989** [reply]
Ok. Thanks for the information. Is there a better way to manage devices? Should the Device service be responsible for deleting the device when it is down?
Also if the device profile is updated, what's the best way to do it so that metadata has the latest profile.

Thanks

**Reply by cloudxxx8** [reply]
The services should call the core-metadata API to do any device and profile CRUD.
You can write the logic you like in your own Device Services.
In normal use cases, devices won't be deleted due to the Device Service shutdown.

**Reply by sounak-1989** [reply]
> The services should call the core-metadata API to do any device and profile CRUD. You can write the logic you like in your own Device Services. In normal use cases, devices won't be deleted due to the Device Service shutdown.

Is there an API in Device SDK that does it or do I need to write code to send HTTP request to core-metadata?

**Reply by cloudxxx8** [reply]
yes, you can retirieve the `DeviceService` instance
https://github.com/edgexfoundry/device-sdk-go/blob/b3f594a181df4350d50f9cefe1b70e34eb008ea7/pkg/service/service.go#L284
and call the related public APIs
https://github.com/edgexfoundry/device-sdk-go/blob/v2.3.1/pkg/service/manageddevices.go
https://github.com/edgexfoundry/device-sdk-go/blob/v2.3.1/pkg/service/managedprofiles.go

**Reply by sounak-1989** [reply]
> yes, you can retirieve the `DeviceService` instance https://github.com/edgexfoundry/device-sdk-go/blob/b3f594a181df4350d50f9cefe1b70e34eb008ea7/pkg/service/service.go#L284 and call the related public APIs https://github.com/edgexfoundry/device-sdk-go/blob/v2.3.1/pkg/service/manageddevices.go https://github.com/edgexfoundry/device-sdk-go/blob/v2.3.1/pkg/service/managedprofiles.go

Thanks for the references.

**Reply by sounak-1989** [reply]
One more query. When I call any API from service.go it takes a while for EdgeX to register. For example if I am setting the operating state of a device to "DOWN" it shows in metadata after a while. The effect is not instant. Any reason why?

Thanks

**Reply by cloudxxx8** [reply]
The effect is instant in my environment.  How do you check the delay?

**Reply by sounak-1989** [reply]
Figured it out. Something on my code. Thanks for the advice.
